src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensEditsTests.cs

```diff
@@ -0,0 +1,307 @@
+                workspace, locations["caret"].First(), SingleLineCase(), previousResultId: "1", cache);
+
+            // 1. Replaces length of token (10 to 5) and replaces token type (comment to keyword)
+            // 2. Creates new token for '// Comment'
+            var expectedEdit = SemanticTokensEditsHandler.GenerateEdit(
+                start: 0, deleteCount: 5,
+                data: new int[]
+                {
+                    0, 0, 5, SemanticTokensCache.TokenTypeToIndex[LSP.SemanticTokenTypes.Keyword], 0,
+                    1, 0, 10, SemanticTokensCache.TokenTypeToIndex[LSP.SemanticTokenTypes.Comment], 0
+                });
+
+            Assert.Equal(expectedEdit, ((LSP.SemanticTokensEdits)results).Edits[0]);
+            Assert.Equal("2", ((LSP.SemanticTokensEdits)results).ResultId);
+        }
+
+        /// <summary>
+        /// Tests to make sure that if we don't have a matching semantic token set for the document in the cache,
+        /// we return the full set of semantic tokens.
+        /// </summary>
+        [Fact]
+        public async Task TestGetSemanticTokensEditsNoCacheAsync()
+        {
+            /*
+             * Original markup:
+             *     // Comment
+             *     static class C { }  
+             */
+            var updatedMarkup =
+@"{|caret:|}// Comment
+
+static class C { }";
+
+            using var workspace = CreateTestWorkspace(updatedMarkup, out var locations);
+            var cache = GetSemanticTokensCache(workspace);
+            var results = await GetSemanticTokensEdits(
+                workspace, locations["caret"].First(), StandardCase(), previousResultId: "10", cache);
+
+            // Make sure we're returned SemanticTokens instead of SemanticTokensEdits.
+            Assert.True(results.Value is LSP.SemanticTokens);
+        }
+
+        [Fact]
+        public async Task TestConvertSemanticTokenEditsIntoSemanticTokens_InsertNewlineInMiddleOfFile()
+        {
+            /*
+             * Original markup:
+             *     // Comment
+             *     static class C { }  
+             */
+            var updatedMarkup =
+@"{|caret:|}// Comment
+
+static class C { }";
+
+            using var workspace = CreateTestWorkspace(updatedMarkup, out var locations);
+            var cache = GetSemanticTokensCache(workspace);
+            var originalTokens = StandardCase();
+
+            // Edits to tokens conversion
+            var edits = await GetSemanticTokensEdits(
+                workspace, locations["caret"].First(), originalTokens, previousResultId: "1", cache);
+            var editsToTokens = ApplySemanticTokensEdits(originalTokens, (LSP.SemanticTokensEdits)edits);
+
+            // Raw tokens
+            var rawTokens = await RunGetSemanticTokensAsync(workspace.CurrentSolution, locations["caret"].First());
+
+            Assert.True(Enumerable.SequenceEqual(rawTokens.Data, editsToTokens));
+        }
```

Comment:
test name is TestConvertSemanticTokenEditsIntoSemanticTokens_InsertNewlineInMiddleOfFile.  but i'm not seeing how a newline was inserted in teh middle of the file.  can you help explain the test for me?

Reply:
The original markup can be found [here](https://github.com/dotnet/roslyn/pull/46407/files#diff-18039c1c8975d6f3dff5f170f34daedbR24) and is:
```C#
// Comment
static class C { }  
```

The updated markup that we test against is:
```C#
// Comment

static class C { }
```

Reply:
yeah. don't love this :-/